src/EffectiveCSharp.Analyzers/PreferDeclarationInitializersToAssignmentStatementsAnalyzer.cs

```diff
@@ -0,0 +1,656 @@
+        {
+            // There are many ways of intializing to zero (0, 0d, 0f, etc), so we use the semantic model to check the constant value
+            // since that is 0 regardless of the syntax used.
+            Optional<object?> constantValue = context.SemanticModel.GetConstantValue(initializer.Value, context.CancellationToken);
+            return constantValue.HasValue && constantValue.Value is int intValue && intValue == 0;
+        }
+
+        // We check if the initializer is a false literal expression.
+        return initializerKind == SyntaxKind.FalseLiteralExpression;
+    }
+
+    /// <summary>
+    /// Determines whether the struct initializer is empty.
+    /// </summary>
+    /// <param name="context">The analysis context.</param>
+    /// <param name="typeSyntax">The type node of the field.</param>
+    /// <param name="initializer">The initializer node to check.</param>
+    /// <returns>A bool on whether the struct initializer is empty.</returns>
+    private static bool IsStructInitializerEmpty(SyntaxNodeAnalysisContext context, TypeSyntax typeSyntax, EqualsValueClauseSyntax initializer)
+    {
+        TypeInfo symbol = context.SemanticModel.GetTypeInfo(typeSyntax, context.CancellationToken);
+
+        // if the type is not a struct or the initializer is not an object creation (meaning the struct is being initialized)
+        // we return false.
+        if (symbol.Type?.TypeKind != TypeKind.Struct
+            || initializer.Value is not ObjectCreationExpressionSyntax objectCreation)
+        {
+            return false;
+        }
+
+        // If the object creation has no arguments, we consider it an empty struct initializer.
+        return objectCreation.ArgumentList?.Arguments.Count == 0;
+    }
+
+    /// <summary>
+    /// Determines whether the field is a zero or null initializable type.
+    /// </summary>
+    /// <param name="fieldSymbol">The field symbol information to check.</param>
+    /// <returns>A bool, on whether the field is a type that can be initialized to null or 0.</returns>
+    private static bool IsZeroOrNullInitializableType(IFieldSymbol fieldSymbol)
+    {
+        return fieldSymbol.Type.IsValueType || fieldSymbol.NullableAnnotation == NullableAnnotation.Annotated;
+    }
+
+    /// <summary>
+    /// Handles fields that are initialized in constructors.
+    /// </summary>
+    /// <param name="context">The <see cref="SyntaxNodeAnalysisContext"/>.</param>
+    /// <param name="field">The <see cref="FieldDeclarationSyntax"/> to handle.</param>
+    /// <param name="initializer">The <see cref="EqualsValueClauseSyntax"/> which contains the field initializer.</param>
+    /// <param name="isInitializerPresent">A bool on whether an initializer is present in the field declaration.</param>
+    /// <param name="fieldInfo">The field's initialization info found in the class's constructors.</param>
+    private static void HandleFieldsInConstructors(
```

Comment:
How does this handle `const`? Constant fields should not be flagged for missing initialization (they are initialized at the point of declaration and cannot be modified). This is an example of a short circuit (e.g. `if (fieldSymbol.IsConst) return;`)

Reply:
I am using the field modifiers to detect when there is a const field, so we don't even bother with going through its declaration. This is better since we avoid the semantic model and just use the declaration